File: ff/Trapping Rain Water/solve.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int trap(vector<int> &height)
    {
        int leng = height.size();
        vector<int> leftLarge(leng);
        leftLarge[0] = height[0];
        for (int i = 1; i < leng; i++)
        {
            leftLarge[i] = max(leftLarge[i - 1], height[i]);
        }
        int rightLarge = height[leng - 1];
        int ans = 0;
        for (int i = leng - 2; i >= 0; i--)
        {
            rightLarge = max(rightLarge, height[i]);
            ans += (min(rightLarge, leftLarge[i]) - height[i]);
        }
        return ans;
    }
};

File: Trapping Rain Water/solve.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function (height) {
    let leng = height.length;
    const leftLarge = new Array(leng);
    leftLarge[0] = height[0];
    for (let i = 1; i < leng; i++) {
        leftLarge[i] = Math.max(leftLarge[i - 1], height[i]);
    }
    let rightLarge = height[leng - 1];
    let ans = 0;
    for (let i = leng - 2; i >= 0; i--) {
        rightLarge = Math.max(rightLarge, height[i]);
        ans += Math.min(rightLarge, leftLarge[i]) - height[i];
    }
    return ans;
};
